chrome/content/zotero/xpcom/server/server_connector.js

```diff
@@ -0,0 +1,1241 @@
+		}
+
+		let { library } = Zotero.Server.Connector.getSaveTarget();
+		if (!library.filesEditable) {
+			return [200, 'text/plain', 'Library files are not editable.'];
+		}
+
+		// We only save the snapshot in single-item cases
+		if (session._action === 'saveSnapshot') {
+			const parentItemID = session.getItemByConnectorKey(data.url).id;
+			// Just saves the snapshot straight up
+			await Zotero.Attachments.importFromSnapshotContent({
+				title: data.title,
+				url: data.url,
+				parentItemID,
+				snapshotContent: data.snapshotContent
+			});
+		}
+		else if (session._action === 'saveItems') {
+			const parentItemID = session.getItemByConnectorKey(data.items[0].id).id;
+			// Deproxifies and does some other attachment preprocessing
+			await session.itemSaver.saveSnapshotAttachments({
+				title: data.title,
+				url: data.url,
+				parentItemID,
+				snapshotContent: data.snapshotContent
+			});
+		}
+
+		return 201;
+	}
+};
+
+/**
+ * Creates a webpage item top-level item in Zotero
+ * Called by the Connector when no translators are detected on the page
+ *
+ * Accepts:
+ *		uri - The URI of the page to be saved
+ *		html - document.innerHTML or equivalent
+ *		cookie - document.cookie or equivalent
+ * Returns:
+ *		Nothing (200 OK response)
+ */
+Zotero.Server.Connector.SaveSnapshot = function() {};
+Zotero.Server.Endpoints["/connector/saveSnapshot"] = Zotero.Server.Connector.SaveSnapshot;
+Zotero.Server.Connector.SaveSnapshot.prototype = {
+	supportedMethods: ["POST"],
+	supportedDataTypes: ["application/json"],
+	permitBookmarklet: true,
+	
+	/**
+	 * Save snapshot
+	 */
+	init: async function (requestData) {
+		var data = requestData.data;
+		
+		var { library, collection } = Zotero.Server.Connector.getSaveTarget();
+		var targetID = collection ? collection.treeViewID : library.treeViewID;
+		
+		try {
+			var session = Zotero.Server.Connector.SessionManager.create(
+				data.sessionID,
+				'saveSnapshot',
+				requestData
+			);
+		}
+		catch (e) {
+			Zotero.debug(e);
+			return [409, "application/json", JSON.stringify({ error: "SESSION_EXISTS" })];
+		}
+		await session.update(collection ? collection.treeViewID : library.treeViewID);
+		
+		// Shouldn't happen as long as My Library exists
+		if (!library.editable) {
+			Zotero.logError("Can't add item to read-only library " + library.name);
+			return [500, "application/json", JSON.stringify({ libraryEditable: false })];
+		}
+		
+		try {
+			await session.saveSnapshot(targetID);
+		}
+		catch (e) {
+			Zotero.logError(e);
+			return 500;
+		}
+		
+		return [201, "application/json"];
+	}
+};
+
+
+/**
+ * Checks if the item has OA attachments (in case PDF saving in connector failed).
+ * Also checks custom resolvers.
+ * 
+ * Accepts:
+ *		sessionID - A session ID previously passed to /saveItems
+ *		itemID - The ID of the item to save alternative attachment for
+ */
+Zotero.Server.Connector.HasOAAttachments = function () {};
```

Comment:
This seems like not the right name. It's not actually checking whether there are OA attachments. It's checking if there are potential resolvers based on the available fields.

Reply:
Yes. `hasAttachmentResolvers`? Should the `getOAAttachments` endpoint change too?

Reply:
`hasAttachmentResolvers` works.

> Should the `getOAAttachments` endpoint change too?

You mean `saveOAAttachment`? Yeah. Um… `saveAttachmentFromResolver`?